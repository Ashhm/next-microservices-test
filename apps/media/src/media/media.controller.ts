import { ConfigurationNamespace } from '@app/shared/common/constants';
import {
  Controller,
  Get,
  HttpException,
  Inject,
  Logger,
  Param,
  Post,
  Response,
  StreamableFile,
  UploadedFile,
  UseInterceptors,
} from '@nestjs/common';
import { ClientProxy } from '@nestjs/microservices';
import { FileInterceptor } from '@nestjs/platform-express';
import { MediaService } from './media.service';

@Controller('media')
export class MediaController {
  private readonly logger = new Logger(MediaController.name);

  constructor(
    private readonly mediaApiService: MediaService,
    @Inject('NETWORK_TEST_SERVICE') private client: ClientProxy,
  ) {}

  @Get('/send-messages')
  async sendMessages() {
    const payload = [1, 2, 3];
    // RPC
    this.client.send<number>({ cmd: 'rpc' }, payload).subscribe();
    // Event
    this.client.emit(ConfigurationNamespace.Application, 'Message body');

    return 'OK';
  }

  @Get('/:filePath')
  async downloadFile(@Param('filePath') filePath: string): Promise<StreamableFile> {
    const stream = await this.mediaApiService.getFileStream(filePath);
    return new Promise((resolve, reject) => {
      stream.on('error', (err) => {
        reject(new HttpException(err.message, 404));
      });
      stream.on('readable', () => {
        stream.removeAllListeners('readable');
        resolve(new StreamableFile(stream, { type: 'image/jpeg' }));
      });
    });
  }

  @Post()
  @UseInterceptors(FileInterceptor('file', { dest: '/mnt/d/tmp' }))
  uploadFile(@UploadedFile() file: Express.Multer.File) {
    return file;
  }
}
