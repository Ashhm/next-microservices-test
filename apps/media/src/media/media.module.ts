import { SharedConfigurationModule } from '@app/shared';
import { NatsQueueName } from '@app/shared/common/constants';
import { NetworkingNatsConfigurationService } from '@app/shared/configuration/networking';
import { Module } from '@nestjs/common';
import { ClientProxyFactory, Transport } from '@nestjs/microservices';
import { MediaController } from './media.controller';
import { MediaService } from './media.service';

@Module({
  imports: [SharedConfigurationModule],
  controllers: [MediaController],
  providers: [
    MediaService,
    {
      provide: 'NETWORK_TEST_SERVICE',
      inject: [NetworkingNatsConfigurationService],
      useFactory: (configService: NetworkingNatsConfigurationService) => {
        return ClientProxyFactory.create({
          transport: Transport.NATS,
          options: {
            servers: [configService.uri],
            queue: NatsQueueName.TestQueue,
          },
          /* transport: Transport.RMQ,
          options: {
            urls: ['amqp://localhost:5672'],
            queue: NatsQueueName.TestQueue,
          },*/
        });
      },
    },
  ],
})
export class MediaModule {}
