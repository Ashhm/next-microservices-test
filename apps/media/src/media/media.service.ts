import * as fs from 'fs';
import { Injectable } from '@nestjs/common';
import * as path from 'path';

@Injectable()
export class MediaService {
  getFileStream(filePath: string): fs.ReadStream {
    return fs.createReadStream(path.join(__dirname, '../../../../', filePath));
  }
}
