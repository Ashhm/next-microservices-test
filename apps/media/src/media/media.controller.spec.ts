import * as path from 'path';
import * as request from 'supertest';
import { NestApplication } from '@nestjs/core';
import { Test, TestingModule } from '@nestjs/testing';
import { MediaController } from './media.controller';
import { MediaService } from './media.service';

describe('MediaController', () => {
  let app: NestApplication;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [MediaController],
      providers: [
        MediaService,
        {
          provide: 'NETWORK_TEST_SERVICE',
          useValue: {
            send: jest.fn(),
            emit: jest.fn(),
          },
        },
      ],
    }).compile();

    app = module.createNestApplication();
    await app.init();
  });

  describe('#downloadFile', () => {
    describe('not existing file', () => {
      it('should return an error', async () => {
        const response = await request(app.getHttpServer()).get('/media/non-existing-file-name');
        expect(response.status).toBe(404);
        expect(response.body).toMatchSnapshot();
      });
    });
    describe('existing file', () => {
      it('should return an error', async () => {
        const response = await request(app.getHttpServer()).get('/media/package.json');
        expect(response.status).toBe(200);
        expect(response.body).toBeInstanceOf(Buffer);
        expect((response.body as Buffer).length).toBeGreaterThan(0);
      });
    });
  });

  describe('#uploadFile', () => {
    it('should return saved file details', async () => {
      const response = await request(app.getHttpServer())
        .post('/media')
        .attach('file', path.join(__dirname, '../../../../', 'package.json'));
      expect(response.status).toBe(201);
      expect(response.body).toMatchSnapshot({
        filename: expect.any(String),
        path: expect.any(String),
        size: expect.any(Number),
      });
    });
  });
});
