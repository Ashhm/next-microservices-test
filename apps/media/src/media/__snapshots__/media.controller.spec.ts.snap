// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaController #downloadFile not existing file should return an error 1`] = `
Object {
  "message": "ENOENT: no such file or directory, open '/mnt/d/tmp/nest-example/non-existing-file-name'",
  "statusCode": 404,
}
`;

exports[`MediaController #uploadFile should return saved file details 1`] = `
Object {
  "destination": "/mnt/d/tmp",
  "encoding": "7bit",
  "fieldname": "file",
  "filename": Any<String>,
  "mimetype": "application/json",
  "originalname": "package.json",
  "path": Any<String>,
  "size": Any<Number>,
}
`;
